### ISSUE TITLE
Checkpointing support

### ISSUE THREAD

**Issue by agerasev** [huggingface/candle#1423]
There is a technique called checkpointing when activations remembered only in specific nodes in computation graph and for other nodes they are recomputed during backpropagation. It allows greatly reduce memory usage at the cost of extra computations, and it is supported in PyTorch: https://pytorch.org/docs/stable/checkpoint.html

I haven't found any functionality related to checkpointing in Candle, so for now I'm using a dirty hack:
+ on feedforward step build an additional computation graph that stores only initial activations,
+ put empty tensors to Candle's backprop graph instead of original tensors to allow them to be removed from memory,
+ use [`CustomOp*`](https://docs.rs/candle-core/latest/candle_core/index.html?search=CustomOp)s to recover removed tensors from additional graph during backprop.

Is there any tool for enabling checkpointing in Candle I haven't found? Or maybe exists a more convenient way to implement it in Candle?

**Comment by agerasev** [comment]
The idea with another graph was not working as expected because we cannot specify an order of Candle's computation graph traversal.

I end up using different approach. Let we have some operation sequence for which we don't want to remember intermediate values. To achieve this the sequence is wrapped in closure and put in `CustomOp*`:

+ On forward pass the closure is executed and result is detached.
+ On backward pass the closure is executed again and `backward` method is called for result, and then we extract gradients for input values from `GradStore`.

This method is not very convenient and requires inner closure to be fully reproducible (e.g. it cannot use random functionality like dropout). But it's pretty simple to implement it without messing with Candle internals.
